src/lib/structures/objective.ml

```diff
@@ -0,0 +1,118 @@
+(*                                                                        *)
+(*     The Alt-Ergo theorem prover                                        *)
+(*                                                                        *)
+(*     Sylvain Conchon, Evelyne Contejean, Francois Bobot                 *)
+(*     Mohamed Iguernelala, Stephane Lescuyer, Alain Mebsout              *)
+(*                                                                        *)
+(*     CNRS - INRIA - Universite Paris Sud                                *)
+(*                                                                        *)
+(*     Until 2013, some parts of this code were released under            *)
+(*     the Apache Software License version 2.0.                           *)
+(*                                                                        *)
+(*     ---------------------------------------------------------------    *)
+(*                                                                        *)
+(*     More details can be found in the directory licenses/               *)
+(*                                                                        *)
+(**************************************************************************)
+
+module Function = struct
+  type t = {
+    e : Expr.t;
+    to_max : bool;
+  }
+
+  let equal { e = e1; _ } { e = e2; _ } = Expr.equal e1 e2
+
+  let mk ~to_max e = { e; to_max }
+
+  let pp ppf { e; _ } = Expr.print ppf e
+end
+
+module Value = struct
+  type limit_kind =
+    | Above
+    | Below
+
+  type t =
+    | Minfinity
+    | Pinfinity
+    | Value of Expr.t
+    | Limit of limit_kind * Expr.t
+    | Unknown
+
+  let pp ppf v =
+    match v with
+    | Minfinity -> Fmt.pf ppf "-oo"
+    | Pinfinity -> Fmt.pf ppf "+oo"
+    | Value w -> Expr.print ppf w
+    | Limit (Above, w) -> Fmt.pf ppf "(+ %a epsilon)" Expr.print w
+    | Limit (Below, w) -> Fmt.pf ppf "(- %a epsilon)" Expr.print w
+    | Unknown -> Fmt.pf ppf "(interval (- oo) (+ oo))"
+end
+
+module Model = struct
+  type s = { f: Function.t; order: int }
+
+  module M =  Map.Make (struct
+      type t = s
+
+      let compare { order = o1; _ } { order = o2; _ } = o1 - o2
+    end)
+
+  type t = Value.t M.t
+
+  let empty = M.empty
+  let is_empty = M.is_empty
+  let fold g = M.fold (fun { f; _ } acc -> g f acc)
+
+  let add o v mdl =
+    let order =
+      match M.find_first (fun { f; _ } -> Function.equal f o) mdl with
+      | ({ order; _ }, _) -> order
+      | exception Not_found -> M.cardinal mdl
+    in
+    M.add { f = o; order } v mdl
+
+  let pp_binding ppf ({ f; _ }, v) =
+    Fmt.pf ppf "(%a %a)" Function.pp f Value.pp v
+
+  let pp ppf mdl =
+    if M.is_empty mdl then
+      Fmt.pf ppf "@[<v 2>(objectives @]@,)"
+    else
+      Fmt.pf ppf "@[<v 2>(objectives @,%a@]@,)"
+        (Fmt.iter_bindings ~sep:Fmt.cut M.iter pp_binding) mdl
+
+  let functions mdl =
+    M.bindings mdl
+    |> List.map (fun ({ f; _ }, _) -> f)
```

Comment:
There probably is some fold function doing this in one go, but I don't expect the map to be too big

Reply:
Actually we have to preserve the order of the objectives. If I traverse my map with a fold or a iter, I will get a list in reversed order and I have to iterate on it a second times to get the correct order.

If we use the reversed order in the map, both `add` and `fold` will be need to reverse the order :/

Reply:
Given that the current code already iterates on the map twice, folding to build a reversed list then calling `List.rev` is actually going to be slightly faster still, but it is probably not significant.